sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp

```diff
@@ -0,0 +1,274 @@
+    typename std::enable_if_t<(LayoutA == matrix::matrix_layout::row_major ||
+                               LayoutA == matrix::matrix_layout::col_major) &&
+                              (LayoutB == matrix::matrix_layout::row_major ||
+                               LayoutB == matrix::matrix_layout::col_major) &&
+                              (LayoutC == matrix::matrix_layout::row_major ||
+                               LayoutC == matrix::matrix_layout::col_major)>> {
+  matrix::joint_matrix<sycl::sub_group, double,
+                       matrix::matrix_type::accumulator, 8, 8, LayoutC>
+  mad(sycl::sub_group sg,
+      matrix::joint_matrix<sycl::sub_group, double, matrix::matrix_type::a, 8,
+                           4, LayoutA>
+          A,
+      matrix::joint_matrix<sycl::sub_group, double, matrix::matrix_type::b, 4,
+                           8, LayoutB>
+          B,
+      matrix::joint_matrix<sycl::sub_group, double,
+                           matrix::matrix_type::accumulator, 8, 8, LayoutC>
+          C) {
+    matrix::joint_matrix<sycl::sub_group, double,
+                         matrix::matrix_type::accumulator, 8, 8, LayoutC>
+        D;
+
+#ifdef __NVPTX__
+#ifdef __SYCL_DEVICE_ONLY__
+    __dmma_m8n8k4_mma_f64(D.data, A.data, B.data, C.data,
+                          get_layout_pair_id<LayoutA, LayoutB>(), 0);
+#endif
+#endif
+
+    return D;
+  }
+};
+
+} // namespace detail
+
+namespace experimental::matrix {
+
+template <typename Group, typename T, matrix_type MT, size_t NumRows,
+          size_t NumCols, matrix_layout Layout, access::address_space Space>
+void joint_matrix_load(
+    Group sg, joint_matrix<Group, T, MT, NumRows, NumCols, Layout> &res,
+    multi_ptr<T, Space> src, size_t stride) {
+  detail::joint_matrix_load_impl<Group, T, MT, NumRows, NumCols, Layout,
+                                 Space>{}
+      .load(res, src, stride);
+}
+
+template <typename Group, typename T, size_t NumRows, size_t NumCols,
+          matrix_layout Layout, access::address_space Space>
+void joint_matrix_store(Group sg,
+                        joint_matrix<Group, T, matrix_type::accumulator,
+                                     NumRows, NumCols, Layout> &src,
+                        multi_ptr<T, Space> dst, size_t stride) {
+  detail::joint_matrix_store_impl<Group, T, NumRows, NumCols, Layout, Space>{}
+      .store(src, dst, stride);
+}
+
+template <typename Group, typename T1, typename T2, std::size_t M,
+          std::size_t K, std::size_t N, matrix_layout LayoutA,
+          matrix_layout LayoutB, matrix_layout LayoutC>
+joint_matrix<Group, T2, matrix_type::accumulator, M, N, LayoutC>
+joint_matrix_mad(
+    Group sg, joint_matrix<Group, T1, matrix_type::a, M, K, LayoutA> A,
+    joint_matrix<Group, T1, matrix_type::b, K, N, LayoutB> B,
+    joint_matrix<Group, T2, matrix_type::accumulator, M, N, LayoutC> C) {
+  return detail::joint_matrix_mad_impl<Group, T1, T2, M, K, N, LayoutA, LayoutB,
```

Comment:
Can you add the file that has the implementation of the detail::load/store/mad functions (new defines_elementary.hpp)?

Reply:
It's here #4695, is that OK?

Reply:
Hi.  matrix-tensorcore.hpp contains the full implementation of the load/store/mad functions which call the correct llvm builtin for each case. Do you mean the implementation of the builtins e.g. __dmma_m8n8k4_ld_a from joint_matrix_load in matrix-tensorcore.hpp?  These builtins are generated by codegen and their declarations are here: https://github.com/intel/llvm/blob/f7c54437772c28d6446dd0801580ec633bfee271/clang/test/CodeGen/builtins-nvptx-mma.cu.  The builtins were implemented via this PR: https://reviews.llvm.org/D104847.